[GTK][WPE] Add more WPE related trace points to better understand the graphics pipeline

nikolaszimmermann opened the pull request:
#### 57b47fb1d205b4dfd9ac47dd5e9aa506f2de7a32
<pre>
[GTK][WPE] Add more WPE related trace points to better understand the graphics pipeline
<a href="https://bugs.webkit.org/show_bug.cgi?id=277412">https://bugs.webkit.org/show_bug.cgi?id=277412</a>

Reviewed by Miguel Gomez.

Add new WPE specific trace points that help us understand the graphics
pipeline, starting from DRM vblank signal (capture by sysprof) until
composition starts/ends for a frame.

Based on a patch from Alejandro G. Castro, refactored to avoid changing
any of the existing enum values - which would be problematic for the Apple
ports.

No new tests - doesn&apos;t affect non-glib ports.

* Source/WTF/wtf/SystemTracing.h:
* Source/WTF/wtf/glib/SysprofAnnotator.h:
* Source/WebKit/Resources/Signposts/SystemTracePoints.plist:
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::renderLayerTree):
* Source/WebKit/UIProcess/DisplayLink.cpp:
(WebKit::DisplayLink::notifyObserversDisplayDidRefresh):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.cpp:
(WebKit::CompositingCoordinator::flushPendingLayerChanges):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::didRenderFrame):

Canonical link: <a href="https://commits.webkit.org/281852@main">https://commits.webkit.org/281852@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b0a01eea7e2f0a8c97a48aaf99e1f561ccc36225

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60734 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40093 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13310 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64665 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11281 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47769 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11556 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49107 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7822 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62768 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37338 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52615 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29939 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34026 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9845 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10194 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/53834 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55866 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10143 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66394 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/59981 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4678 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9975 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56475 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4699 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52587 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56654 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13648 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3878 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/81736 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35898 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14227 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36980 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38073 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36725 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WTF/wtf/SystemTracing.h:
```diff
@@ -163,10 +163,21 @@ enum TracePointCode {
     WebXRCPFrameStartSubmissionEnd,
     WebXRCPFrameEndSubmissionStart,
     WebXRCPFrameEndSubmissionEnd,
+    DisplayLinkUpdate,
```
This is in WebCore group, but it's added in the UI process, I guess it should be in WebKit2 group instead.

nikolaszimmermann replied:
Good point, will fix.

carlosgcampos commented on Source/WTF/wtf/SystemTracing.h:
```diff
@@ -163,10 +163,21 @@ enum TracePointCode {
     WakeUpAndApplyDisplayListEnd,
+
+#if USE(GLIB)
+    GLibPortRange = 20000,
```
I think this is a bit confusing, because there's no GLib port and because these are not glib specific events. Maybe we could add a CompositionRange or something like that.

nikolaszimmermann replied:
They are all gtk/wpe specific, and only used for these ports. That's why I went with USE(GLIB) instead of PLATFORM(GTK) || PLATFORM(WPE).

The only addition here in this patch that is cross-platform is 'DisplayLinkUpdate'.

carlosgcampos commented on Source/WTF/wtf/SystemTracing.h:
```diff
@@ -163,10 +163,21 @@ enum TracePointCode {
+    GLibPortRange = 20000,
+
+    DidRenderFrameStart,
+    DidRenderFrameEnd,
```
I see this is named like the method where this is added, but that's already confusing. For the trace name I think it's better to use a more accurate name like `DidComposeFrameStart/End`

nikolaszimmermann replied:
Makes sense. I left the names from @alexgcastro variant before, but let's get it right now.

nikolaszimmermann replied:
Went for 'WaitForCompositionCompletion' (since either we don't wait here in case of explicit sync, or clientWait() on the fence).

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.cpp:
```diff
@@ -161,6 +162,7 @@ void CompositingCoordinator::sizeDidChange(const IntSize& newSize)
 
 bool CompositingCoordinator::flushPendingLayerChanges(OptionSet<FinalizeRenderingUpdateFlags> flags)
 {
+    TraceScope traceScope(BackingStoreFlushStart, BackingStoreFlushEnd);
```
Is this really equivalent to BackingStoreFlush? Again, I would use a more accurate name like `LayerFlushStart/End`

nikolaszimmermann replied:
It's not, since we now have our own category, let's add instead of reusing things.

GeorgesStavracas commented on Source/WTF/wtf/SystemTracing.h:
```diff
@@ -163,10 +163,21 @@ enum TracePointCode {
     WebXRCPFrameStartSubmissionEnd,
     WebXRCPFrameEndSubmissionStart,
     WebXRCPFrameEndSubmissionEnd,
+    DisplayLinkUpdate,
```
Right above this enum, there's a comment that reads:

> // When adding or changing these codes, update Tools/Tracing/SystemTracePoints.plist to match.

This file seems to still exist, so I think this should be done.

FWIW, the reason I suggested not using this enum / TraceScope was specifically to avoid editing the plist file. I can see how TraceScope is massively useful though!

Maybe it's worth double checking with Apple folks if just editing the plist file is enough.

nikolaszimmermann replied:
Forgot about the plist file, thanks for mentioning. The only thing that needs to be added there would be 'DisplayLinkUpdate', since the other new tracing points are GTK/WPE specific.

I remember from the past: we shall not add enum entries in the middle, but append.
I do want to double-check with Apple. @smfr, maybe?

nikolaszimmermann replied:
In fact ` Tools/Tracing/SystemTracePoints.plist` is probably the reason why we cannot change enum values "in the middle", but appending works. I've added a new entry for 'DisplayLinkUpdate', but not for the GTK/WPE specific ones.

GeorgesStavracas reviewed (commented):
One comment. I think I agree with all of @carlosgcampos suggestions. Overall, these new tracing points look very useful!

GeorgesStavracas reviewed (approved):
LGTM now!

nikolaszimmermann committed:
[GTK][WPE] Add more WPE related trace points to better understand the graphics pipeline
https://bugs.webkit.org/show_bug.cgi?id=277412

Reviewed by Miguel Gomez.

Add new WPE specific trace points that help us understand the graphics
pipeline, starting from DRM vblank signal (capture by sysprof) until
composition starts/ends for a frame.

Based on a patch from Alejandro G. Castro, refactored to avoid changing
any of the existing enum values - which would be problematic for the Apple
ports.

No new tests - doesn't affect non-glib ports.

* Source/WTF/wtf/SystemTracing.h:
* Source/WTF/wtf/glib/SysprofAnnotator.h:
* Source/WebKit/Resources/Signposts/SystemTracePoints.plist:
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::renderLayerTree):
* Source/WebKit/UIProcess/DisplayLink.cpp:
(WebKit::DisplayLink::notifyObserversDisplayDidRefresh):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.cpp:
(WebKit::CompositingCoordinator::flushPendingLayerChanges):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::didRenderFrame):

Canonical link: https://commits.webkit.org/281852@main

rkirsling commented on Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:
```diff
@@ -220,6 +221,8 @@ void ThreadedCompositor::forceRepaint()
 
 void ThreadedCompositor::renderLayerTree()
 {
+    TraceScope traceScope(FrameCompositionStart, FrameCompositionEnd);
```
This line isn't in a GTK/WPE-specific context, added a guard in #31759.

nikolaszimmermann replied:
Thanks @rkirsling. sorry for the inconveniences, will remember next time the compositor is used also for other ports.
